[reland] Support torch.distributed.irecv(src=None, ...)

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49383 [reland] Support torch.distributed.irecv(src=None, ...)**

Reland of https://github.com/pytorch/pytorch/pull/47137

Differential Revision: [D25551910](https://our.internmc.facebook.com/intern/diff/D25551910/)

pritamdamania committed:
[reland] Support torch.distributed.irecv(src=None, ...)

Reland of https://github.com/pytorch/pytorch/pull/47137

Differential Revision: [D25551910](https://our.internmc.facebook.com/intern/diff/D25551910/)

[ghstack-poisoned]

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -855,27 +855,56 @@ def test_send_recv(self):
+                    dist.send(tensor, dst)  # recv
+                    dist.send(tensor, dst)  # irecv
+
+            # Each rank would have 2 * (world_size - 1) sends, verify that
+            # globally we receive the same amount on the other end.
+            recv_ranks_tensor = torch.cat((torch.tensor(recv_ranks), torch.tensor(irecv_ranks)), 0)
+            global_recv_ranks = [
+                torch.empty_like(recv_ranks_tensor),
+                torch.empty_like(recv_ranks_tensor),
+                torch.empty_like(recv_ranks_tensor),
+                torch.empty_like(recv_ranks_tensor),
+            ]
+            dist.all_gather(global_recv_ranks, recv_ranks_tensor)
+            global_recv_ranks_list = []
+            for tensor in global_recv_ranks:
+                global_recv_ranks_list += tensor.tolist()
+
+            from itertools import groupby
+            global_recv_ranks_list.sort()
+            frequency = [len(list(group)) for key, group in groupby(global_recv_ranks_list)]
+            self.assertEqual(dist.get_world_size(), len(frequency))
+            self.assertEqual([2 * (dist.get_world_size() - 1)] * dist.get_world_size(), frequency)
```
The validation here has been made more robust compared to #47137, since recvAnySource can potentially recv from anywhere.

pritamdamania committed:
Update on "[reland] Support torch.distributed.irecv(src=None, ...)"


Reland of https://github.com/pytorch/pytorch/pull/47137

Differential Revision: [D25551910](https://our.internmc.facebook.com/intern/diff/D25551910/)

[ghstack-poisoned]

osalpekar commented on torch/distributed/distributed_c10d.py:
```diff
@@ -718,11 +719,18 @@ def irecv(tensor,
+        return pg.recv_anysource([tensor], tag)
+    else:
+        if pg is GroupMember.WORLD:
+            pg.recv([tensor], src, tag).wait()
```
Why do we block in this case?

pritamdamania87 replied:
Good catch, this is a bug.

pritamdamania committed:
Update on "[reland] Support torch.distributed.irecv(src=None, ...)"


Reland of https://github.com/pytorch/pytorch/pull/47137

Differential Revision: [D25551910](https://our.internmc.facebook.com/intern/diff/D25551910/)

[ghstack-poisoned]

pritamdamania committed:
Update on "[reland] Support torch.distributed.irecv(src=None, ...)"


Reland of https://github.com/pytorch/pytorch/pull/47137

Differential Revision: [D25551910](https://our.internmc.facebook.com/intern/diff/D25551910/)

[ghstack-poisoned]
